pkg/apis/apps/v1/defaults.go

```diff
@@ -103,8 +103,7 @@ func SetDefaults_StatefulSet(obj *appsv1.StatefulSet) {
 		obj.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateStatefulSetStrategy{}
 	}
 
-	if obj.Spec.UpdateStrategy.Type == appsv1.RollingUpdateStatefulSetStrategyType &&
-		obj.Spec.UpdateStrategy.RollingUpdate != nil &&
+	if obj.Spec.UpdateStrategy.RollingUpdate != nil &&
 		obj.Spec.UpdateStrategy.RollingUpdate.Partition == nil {
 		obj.Spec.UpdateStrategy.RollingUpdate.Partition = new(int32)
 		*obj.Spec.UpdateStrategy.RollingUpdate.Partition = 0
```

Comment:
`spec.updateStrategy.rollingUpdate` should be unset if `spec.updateStartegy` isn't `RollingUpdate`, ref #35345

Is it not possible to leave it unset?

Reply:
> `spec.updateStrategy.rollingUpdate` should be unset if `spec.updateStartegy` isn't `RollingUpdate`

yes, this rule is enforced during validation. This PR changes the following behavior:

If `spec.updateStrategy.rollingUpdate` is set and `spec.updateStartegy` isn't `RollingUpdate`, e.g. 
```
  updateStrategy:
    type: OnDelete
    rollingUpdate: {}
```

- before this change, apiserver panics during conversion and the client sees a 500 internal server error
- after this change, apiserver rejects during validation and the client sees the validation error

ref https://github.com/kubernetes/kubernetes/issues/81616 for more info. Alternatively we can fix the conversion function, but changing the defaulting seems more aligned with the godoc

Reply:
> Alternatively we can fix the conversion function, but changing the defaulting seems more aligned with the godoc

IMO current defaulting behavior is more aligned with godoc, which states that `rollingUpdate` is only used by the type is `RollingUpdate`:

```
// RollingUpdate is used to communicate parameters when Type is RollingUpdateStatefulSetStrategyType.
```

Note that changing defaults is breaking. This would trigger a (surprising and unwanted) rolling update to users' StatefulSet when they upgrade their clusters. 

It'd be safer & compatible fixing conversion instead. Also, would you add a test to catch the API server panic issue too? Thanks!

Reply:
@janetkuo Thanks for the suggestion! It turns out that the conversion has been fixed by 
https://github.com/kubernetes/kubernetes/pull/82780. The current conversion code (auto-generated) handles `*int -> int` conversion properly:
https://github.com/kubernetes/kubernetes/blob/5e31799701123c50025567b8534e1a62dbc0e9f6/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/conversion.go#L49-L52

I tested locally and the panic issue is gone

/close